provider/azure/environ.go

```diff
@@ -1604,25 +1562,170 @@ func (env *azureEnviron) updateResourceControllerTag(
 	return errorutils.HandleCredentialError(errors.Annotatef(err, "updating controller for %q", to.String(resource.Name)), ctx)
 }
 
+var (
+	runningInstStates = []compute.ProvisioningState{
+		compute.ProvisioningStateCreating,
+		compute.ProvisioningStateUpdating,
+		compute.ProvisioningStateMigrating,
+		compute.ProvisioningStateSucceeded,
+	}
+)
+
+// Instances is specified in the Environ interface.
+func (env *azureEnviron) Instances(ctx context.ProviderCallContext, ids []instance.Id) ([]instances.Instance, error) {
+	if len(ids) == 0 {
+		return nil, nil
+	}
+	insts := make([]instances.Instance, len(ids))
+	// Make a series of requests to cope with eventual consistency.
+	// Each request will attempt to add more instances to the requested set.
+	err := retry.Call(retry.CallArgs{
+		Func: func() error {
+			var need []instance.Id
+			for i, inst := range insts {
+				if inst == nil {
+					need = append(need, ids[i])
+				}
+			}
+			return env.gatherInstances(ctx, need, insts, env.resourceGroup, true)
+		},
+		IsFatalError: func(err error) bool {
+			return err != environs.ErrPartialInstances
+		},
+		Attempts:    -1,
+		Delay:       200 * time.Millisecond,
+		MaxDuration: 5 * time.Second,
+		Clock:       env.provider.config.RetryClock,
+	})
+
+	if err == environs.ErrPartialInstances {
+		for _, inst := range insts {
+			if inst != nil {
+				return insts, environs.ErrPartialInstances
+			}
+		}
+		return nil, environs.ErrNoInstances
+	}
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	return insts, nil
+}
+
 // AllInstances is specified in the InstanceBroker interface.
 func (env *azureEnviron) AllInstances(ctx context.ProviderCallContext) ([]instances.Instance, error) {
-	return env.allInstances(ctx, env.resourceGroup, true /* refresh addresses */, false /* all instances */)
+	return env.allInstances(ctx, env.resourceGroup, true, "")
 }
 
 // AllRunningInstances is specified in the InstanceBroker interface.
 func (env *azureEnviron) AllRunningInstances(ctx context.ProviderCallContext) ([]instances.Instance, error) {
-	return env.AllInstances(ctx)
+	return env.allInstances(ctx, env.resourceGroup, true, "", runningInstStates...)
+}
+
+// gatherInstances tries to get information on each instance id
+// whose corresponding insts slot is nil.
+// This function returns environs.ErrPartialInstances if the
+// insts slice has not been completely filled.
+func (env *azureEnviron) gatherInstances(
```

Comment:
Azure doesn't have an API for querying instances by a slice of IDs?

Reply:
sadly no - you can ask for instances in a given resource group, but the api has no filter parameter